Source/BugsnagMetadata.m

```diff
@@ -43,81 +44,52 @@ - (id)init {
 
 - (id)initWithDictionary:(NSMutableDictionary *)dict {
     if (self = [super init]) {
-        self.dictionary = dict;
+        // Ensure that the instantiating dictionary is mutable.
+        // Saves checks later.
+        self.dictionary = [dict mutableDeepCopy];
     }
     [self.delegate metadataChanged:self];
     return self;
 }
 
-// MARK: - <NSMutableCopying>
-
-- (id)mutableCopyWithZone:(NSZone *)zone {
-    @synchronized(self) {
-        NSMutableDictionary *dict = [self.dictionary mutableCopy];
-        return [[BugsnagMetadata alloc] initWithDictionary:dict];
-    }
-}
-
-- (NSMutableDictionary *)getMetadata:(NSString *)sectionName {
-    @synchronized(self) {
-        return self.dictionary[sectionName];
-    }
-}
-
-- (NSMutableDictionary *)getMetadata:(NSString *)sectionName
-                                 key:(NSString *)key
+- (NSDictionary *)toDictionary
 {
     @synchronized(self) {
-        return [self.dictionary valueForKeyPath:[NSString stringWithFormat:@"%@.%@", sectionName, key]];
-    }
-}
-
-- (void)clearMetadataInSection:(NSString *)sectionName {
-    @synchronized(self) {
-        [self.dictionary removeObjectForKey:sectionName];
+        return [NSDictionary dictionaryWithDictionary:self.dictionary];
     }
-    [self.delegate metadataChanged:self];
 }
 
-- (void)clearMetadataInSection:(NSString *)section
-                           key:(NSString *)key
-{
+- (id)deepCopy {
     @synchronized(self) {
-        if ([[[self dictionary] objectForKey:section] objectForKey:key]) {
-            [[[self dictionary] objectForKey:section] removeObjectForKey:key];
-        }
+        return [[BugsnagMetadata alloc] initWithDictionary:[[self dictionary] mutableDeepCopy]];
```

Comment:
nit: the `mutableDeepCopy` can be avoided here as the constructor implementation already makes a defensive copy